x-pack/apm-server/sampling/pubsub/pubsub.go

```diff
@@ -0,0 +1,219 @@
+//
+// Documents are expected to be indexed through a pipeline which sets the
+// `event.ingested` timestamp field. Another process will periodically reap
+// events older than a configured age.
+func New(config Config) (*Pubsub, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid pubsub config")
+	}
+	if config.Logger == nil {
+		config.Logger = logp.NewLogger(logs.Sampling)
+	}
+	indexer, err := esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
+		Client:        config.Client,
+		Index:         config.Index,
+		FlushInterval: config.FlushInterval,
+		OnError: func(ctx context.Context, err error) {
+			config.Logger.With(logp.Error(err)).Debug("publishing sampled trace IDs failed")
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Pubsub{config: config, indexer: indexer}, nil
+}
+
+// PublishSampledTraceIDs bulk indexes traceIDs into Elasticsearch.
+func (p *Pubsub) PublishSampledTraceIDs(ctx context.Context, traceID ...string) error {
+	for _, id := range traceID {
+		var doc traceIDDocument
+		doc.Observer.ID = p.config.BeatID
+		doc.Trace.ID = id
+
+		var json fastjson.Writer
+		if err := doc.MarshalFastJSON(&json); err != nil {
+			return err
+		}
+		if err := p.indexer.Add(ctx, esutil.BulkIndexerItem{
+			Action: "index",
+			Body:   bytes.NewReader(json.Bytes()),
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SubscribeSampledTraceIDs subscribes to new sampled trace IDs, sending them to the
+// traceIDs channel.
+func (p *Pubsub) SubscribeSampledTraceIDs(ctx context.Context, traceIDs chan<- string) error {
+	ticker := time.NewTicker(p.config.SearchInterval)
+	defer ticker.Stop()
+
+	// NOTE(axw) we should use the Changes API when it is implemented:
+	// https://github.com/elastic/elasticsearch/issues/1242
+
+	var lastSeqNo int64 = 0
+	var lastPrimaryTerm int64 = -1
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+		}
+		for {
+			// Keep searching until there are no more new trace IDs.
+			n, err := p.searchTraceIDs(ctx, traceIDs, &lastSeqNo, &lastPrimaryTerm)
+			if err != nil {
+				// Errors may occur due to rate limiting, or while the index is
+				// still being created, so just log and continue.
+				p.config.Logger.With(logp.Error(err)).Debug("error searching for trace IDs")
+				break
+			}
+			if n == 0 {
+				// No more results, go back to sleep.
+				break
+			}
+		}
+	}
+}
+
+// searchTraceIDs searches for new sampled trace IDs (after lastPrimaryTerm and lastSeqNo),
+// sending them to the out channel and returning the number of trace IDs sent.
+func (p *Pubsub) searchTraceIDs(ctx context.Context, out chan<- string, lastSeqNo, lastPrimaryTerm *int64) (int, error) {
+	searchBody := map[string]interface{}{
+		"size":                1000,
+		"seq_no_primary_term": true,
+		"sort":                []interface{}{map[string]interface{}{"_seq_no": "asc"}},
+		"search_after":        []interface{}{*lastSeqNo - 1},
```

Comment:
Could you add a comment on why you need to do the `-1` here; took me some time to figure out it's because of the potentially increased `_primary_term`.